### PR TITLE
Feature/manage map image templates

### PR DESCRIPTION
**Pull request by Dassderdie**
#260 

image

image

image

### CODE REVIEW

**Review by hpistudent72 — DISMISSED** [review]
I believe while editinig a template, you should also be able to change the url.

Looks good otherwise

File: shared/src/store/action-reducers/map-image-templates.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+export class AddMapImageTemplateAction implements Action {
+    @IsString()
+    public readonly type = '[MapImageTemplate] Add mapImageTemplate';
+
+    @ValidateNested()
+    @Type(() => MapImageTemplate)
+    public readonly mapImageTemplate!: MapImageTemplate;
+}
+
+export class EditMapImageTemplateAction implements Action {
```
Comment (ClFeSc) [inline shared/src/store/action-reducers/map-image-templates.ts:25]: What about editing the URL?

File: frontend/src/app/pages/exercises/exercise/shared/trainer-map-editor/trainer-map-editor.component.html [hunk 2]
```diff
@@ -167,7 +167,16 @@ <h5>Fahrzeuge</h5>
                     </div>
                 </li>
                 <li class="list-group-item">
-                    <h5>Bilder</h5>
+                    <h5>
```
Comment (ClFeSc) [inline frontend/src/app/pages/exercises/exercise/shared/trainer-map-editor/trainer-map-editor.component.html:170]: Maybe add a text similar to the transfer overview when no image template exists. It looks a bit weird right now, in my opinion.

**Review by ClFeSc — CHANGES_REQUESTED** [review]

File: frontend/src/app/pages/exercises/exercise/shared/editor-panel/image-template-form/image-template-form.component.html [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        [(ngModel)]="imageTemplate.name"
+        required
+        type="text"
+        class="form-control"
+        placeholder="Bildname"
+    />
+    <app-display-validation [ngModelInput]="nameInput"></app-display-validation>
+</div>
+<div class="form-group mb-3">
+    <label class="form-label">Höhe:</label>
```
[the remark this answers is not in the document]
Comment (Dassderdie) [reply frontend/src/app/pages/exercises/exercise/shared/editor-panel/image-template-form/image-template-form.component.html:30]: We will convert this to a meaningful unit later. Currently, the unit is not understandable to anyone.

File: frontend/src/app/pages/exercises/exercise/shared/editor-panel/create-image-template-modal/create-image-template-modal.component.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+@Component({
+    selector: 'app-create-image-template-modal',
+    templateUrl: './create-image-template-modal.component.html',
+    styleUrls: ['./create-image-template-modal.component.scss'],
+})
+export class CreateImageTemplateModalComponent implements OnDestroy {
+    @Output() readonly createImageTemplate$ =
+        new EventEmitter<MapImageTemplate | null>();
+
+    public readonly imageTemplate: EditableImageTemplate = {
```
[the remark this answers is not in the document]
Comment (ClFeSc) [reply frontend/src/app/pages/exercises/exercise/shared/editor-panel/create-image-template-modal/create-image-template-modal.component.ts:19]: Ah, ok. Then leave it as it is.

File: frontend/src/app/pages/exercises/exercise/shared/editor-panel/image-template-form/image-template-form.component.html (hunk 3 above)
[the remark this answers is not in the document]
Comment (ClFeSc) [reply frontend/src/app/pages/exercises/exercise/shared/editor-panel/image-template-form/image-template-form.component.html:30]: See #374

File: frontend/src/app/pages/exercises/exercise/shared/editor-panel/create-image-template-modal/create-image-template-modal.component.ts [hunk 5]
```diff
@@ -32,22 +32,21 @@ export class CreateImageTemplateModalComponent implements OnDestroy {
     ) {}
 
     public async createImageTemplate() {
```
Comment (ClFeSc) [inline frontend/src/app/pages/exercises/exercise/shared/editor-panel/create-image-template-modal/create-image-template-modal.component.ts:34]:
```suggestion
    public createImageTemplate() {
```

I don't think this is needed anymore.

File: frontend/src/app/shared/functions/get-image-aspect-ratio.ts [hunk 6]
```diff
@@ -1,3 +1,7 @@
+/**
+ * @returns wether the a Promise that resolves with the aspect ratio of the image
+ * it rejects if the image Url is invalid
```
Comment (ClFeSc) [inline frontend/src/app/shared/functions/get-image-aspect-ratio.ts:3]: This comment is a bit messed up.

File: frontend/src/app/shared/functions/get-image-aspect-ratio.ts [hunk 7]
```diff
@@ -1,3 +1,7 @@
...
+ */
```
Comment (ClFeSc) [inline frontend/src/app/shared/functions/get-image-aspect-ratio.ts:4]: I don't know why, but this doesn't work. It rejects for URLs that exist. You can test this with any remote URL, e.g. `https://hpi.de/typo3conf/ext/hpi_distribution/Resources/Public/Default/img/hpi_logo_srgb_wb_sl1_web80.png` (a low-res HPI logo). It exists but isn't recognized. The error is `Failed to fetch`.

**Review by ClFeSc — CHANGES_REQUESTED** [review]
Sadly I noticed some more issues.

Also, the aspect ratio of edited images is incorrect. Just e.g. change `house-fire.svg` to `fire.svg` and observe what happens when the image is dragged onto the map.

**Comment by ClFeSc** [comment]
Another thought that just crossed my mind: What happens when you drag a template while another trainer deletes it and you then drop it?

File: frontend/src/app/pages/exercises/exercise/shared/editor-panel/image-template-form/image-template-form.component.html [hunk 8]
```diff
@@ -1,53 +1,57 @@
...
+            class="form-control"
+            placeholder="Bildname"
+        />
+        <app-display-validation
+            [ngModelInput]="nameInput"
+        ></app-display-validation>
+    </div>
+    <div class="form-group mb-3">
+        <label class="form-label">Höhe:</label>
+        <input
```
Comment (ClFeSc) [inline frontend/src/app/pages/exercises/exercise/shared/editor-panel/image-template-form/image-template-form.component.html:34]: This input permits `0` which then fails the proposal.

File: frontend/src/app/pages/exercises/exercise/shared/editor-panel/create-image-template-modal/create-image-template-modal.component.ts (hunk 5 above)
Comment (ClFeSc) [reply frontend/src/app/pages/exercises/exercise/shared/editor-panel/create-image-template-modal/create-image-template-modal.component.ts]: Thumbs up but no change?

File: frontend/src/app/pages/exercises/exercise/shared/editor-panel/image-template-form/image-template-form.component.html [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
```
Comment (ClFeSc) [inline frontend/src/app/pages/exercises/exercise/shared/editor-panel/image-template-form/image-template-form.component.html:34]: Sorry, but I'd like to have the input rejected before proposing anything if the height is not an integer.

**Review by ClFeSc — CHANGES_REQUESTED** [review]

File: frontend/src/app/pages/exercises/exercise/shared/editor-panel/image-template-form/image-template-form.component.html (hunk 9 above)
Comment (Dassderdie) [reply frontend/src/app/pages/exercises/exercise/shared/editor-panel/image-template-form/image-template-form.component.html:34]: Well, I don't really think the cost/benefit justifies this feature, but ok I'll do it...

**Pull request merged**